contracts/test/FeeContract.t.sol

```diff
@@ -0,0 +1,278 @@
+        amount = bound(
+            amount, feeContractProxy.MIN_DEPOSIT_AMOUNT(), feeContractProxy.MAX_DEPOSIT_AMOUNT()
+        );
+
+        uint256 balanceBefore = feeContractProxy.getBalance(user);
+
+        //deposit for the user
+        feeContractProxy.deposit{ value: amount }(user);
+
+        //get the balance for that user after the deposit
+        uint256 balanceAfter = feeContractProxy.getBalance(user);
+        assertEq(balanceAfter, balanceBefore + amount);
+    }
+
+    //test that depositing twice increases the user's baalance
+    function testFuzz_depositTwice(address user) public payable {
+        if (msg.value == 0) return;
+        vm.prank(user);
+        uint256 balanceBefore = feeContractProxy.getBalance(user);
+
+        uint256 depositAmount = msg.value / 2;
+        //deposit for the user
+        feeContractProxy.deposit{ value: depositAmount }(user);
+
+        //get the balance for that user after the deposit
+        uint256 balanceAfter = feeContractProxy.getBalance(user);
+        assertEq(balanceAfter, balanceBefore + depositAmount);
+
+        //deposit the remainder for the user
+        feeContractProxy.deposit{ value: depositAmount }(user);
+
+        //get the balance for that user after the 2nd deposit
+        uint256 balanceAfter2 = feeContractProxy.getBalance(user);
+        assertEq(balanceAfter2, balanceAfter + depositAmount);
+    }
+
+    function testFuzz_noFunction() public payable {
+        address fcAddress = address(feeContractProxy);
+        (bool success,) = fcAddress.call{ value: msg.value }("");
+
+        //assert that the transaction was not successful
+        assertFalse(success);
+
+        //assert that the balance of the fee contract is still zero
+        assertEq(address(feeContractProxy).balance, 0);
+    }
+
+    function testFuzz_nonExistentFunction() public payable {
+        address fcAddress = address(feeContractProxy);
+        (bool success,) =
+            fcAddress.call{ value: msg.value }(abi.encodeWithSignature("withdraw(address)", "0x"));
+
+        //assert that the transaction was not successful
+        assertFalse(success);
+
+        //assert that the balance of the fee contract is still zero
+        assertEq(address(feeContractProxy).balance, 0);
+    }
+
+    function testUpgradeTo() public {
+        FeeContractV2Test feeContractV2 = new FeeContractV2Test();
+
+        vm.prank(admin);
+        vm.expectEmit(false, false, false, true);
+        // We emit the event we expect to see.
+        emit FeeContract.Upgrade(address(feeContractV2));
+
+        feeContractProxy.upgradeToAndCall(address(feeContractV2), "");
+    }
+
+    function testFailUpgradeToWithWrongAdmin() public {
+        FeeContractV2Test feeContractV2 = new FeeContractV2Test();
+
+        //start the upgrade with another address which isn't the admin
+        vm.prank(makeAddr("otherUser"));
+        feeContractProxy.upgradeToAndCall(address(feeContractV2), "");
```

Comment:
we should add `expectRevert` ?

also should we test that if the we can upgrade to a `V2` with no admin? and/or become non-upgradable?

Reply:
> also should we test that if the we can upgrade to a V2 with no admin? and/or become non-upgradable?
so test whether it's possible to upgrade to another contract that isn't upgradeable?

Reply:
> we should add expectRevert ?
I used the `testFail` syntax which means that the foundry test fails when the contract call reverts. Currently checking with the team to see if there is a preferred testing method for expected failures.

Reply:
> > we should add expectRevert ?
> 
> I used the `testFail` syntax which means that the foundry test fails when the contract call reverts. Currently checking with the team to see if there is a preferred testing method for expected failures.

since we decided to go with expectRevert, I've updated the test https://github.com/EspressoSystems/espresso-sequencer/commit/d4e477df94a8b75a1dd3ec1a30d75363afc5fb4d